File: Packages/mcp-unity/Server~/src/tools/prefabCreateFromSceneTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'prefab_create_from_scene';
const toolDescription = 'Creates a prefab asset from a scene GameObject by instance ID or hierarchy path.';

const paramsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject'),
  objectPath: z.string().optional().describe('The hierarchy path of the GameObject'),
  prefabPath: z.string().describe("Destination prefab asset path (e.g. 'Assets/Prefabs/MyPrefab.prefab')")
});

export function registerPrefabCreateFromSceneTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: z.infer<typeof paramsSchema>) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>): Promise<CallToolResult> {
  const { instanceId, objectPath, prefabPath } = params;

  if (instanceId === undefined && (!objectPath || objectPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'instanceId' or 'objectPath' must be provided"
    );
  }

  if (!prefabPath || prefabPath.trim() === '') {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'prefabPath' not provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: { instanceId, objectPath, prefabPath }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to create prefab from scene object'
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message || 'Successfully created prefab asset'
    }],
    data: {
      prefabPath: response.prefabPath,
      guid: response.guid,
      instanceId: response.instanceId
    }
  };
}

File: Packages/mcp-unity/Editor/Tools/PrefabCreateFromSceneTool.cs
using System;
using System.IO;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for creating a prefab asset from a scene GameObject.
    /// </summary>
    public class PrefabCreateFromSceneTool : McpToolBase
    {
        public PrefabCreateFromSceneTool()
        {
            Name = "prefab_create_from_scene";
            Description = "Creates a prefab asset from a scene GameObject by instance ID or hierarchy path.";
        }

        public override JObject Execute(JObject parameters)
        {
            int? instanceId = parameters?["instanceId"]?.ToObject<int?>();
            string objectPath = parameters?["objectPath"]?.ToObject<string>();
            string prefabPath = parameters?["prefabPath"]?.ToObject<string>();

            if (!instanceId.HasValue && string.IsNullOrEmpty(objectPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Either 'instanceId' or 'objectPath' must be provided",
                    "validation_error"
                );
            }

            if (string.IsNullOrEmpty(prefabPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'prefabPath' not provided",
                    "validation_error"
                );
            }

            prefabPath = NormalizeAssetPath(prefabPath.Trim());

            if (!prefabPath.StartsWith("Assets", StringComparison.Ordinal) ||
                !prefabPath.EndsWith(".prefab", StringComparison.OrdinalIgnoreCase))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Parameter 'prefabPath' must start with 'Assets' and end with '.prefab'",
                    "validation_error"
                );
            }

            JObject error = GameObjectToolUtils.FindGameObject(instanceId, objectPath, out GameObject gameObject, out _);
            if (error != null)
            {
                return error;
            }

            EnsureFolderHierarchy(Path.GetDirectoryName(prefabPath));

            PrefabUtility.SaveAsPrefabAsset(gameObject, prefabPath, out bool prefabSaved);
            if (!prefabSaved)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Failed to create prefab asset at '{prefabPath}'",
                    "prefab_save_error"
                );
            }

            string guid = AssetDatabase.AssetPathToGUID(prefabPath);

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Successfully created prefab asset at '{prefabPath}'",
                ["prefabPath"] = prefabPath,
                ["guid"] = guid,
                ["instanceId"] = McpObjectId.FromObject(gameObject)
            };
        }

        private static void EnsureFolderHierarchy(string folderPath)
        {
            folderPath = NormalizeAssetPath(folderPath);

            if (string.IsNullOrEmpty(folderPath) || AssetDatabase.IsValidFolder(folderPath))
            {
                return;
            }

            string parentPath = NormalizeAssetPath(Path.GetDirectoryName(folderPath));
            string folderName = Path.GetFileName(folderPath);

            if (!string.IsNullOrEmpty(parentPath) && !AssetDatabase.IsValidFolder(parentPath))
            {
                EnsureFolderHierarchy(parentPath);
            }

            if (!AssetDatabase.IsValidFolder(folderPath) && !string.IsNullOrEmpty(parentPath) && !string.IsNullOrEmpty(folderName))
            {
                AssetDatabase.CreateFolder(parentPath, folderName);
            }
        }

        private static string NormalizeAssetPath(string path)
        {
            return string.IsNullOrEmpty(path) ? path : path.Replace('\\', '/');
        }
    }
}
